Remove Python 3.8 and 3.9, add Python 3.11

mary-cleaton opened the pull request:
## Description

Remove Python 3.8 and 3.9 (as they are unsupported) and add Python 3.11 (as it is the most recent version of Python that works with Spark 3.5.x).

closes #51

## Type of change

- [ ] Bug fix - *non-breaking change*
- [ ] New feature - *non-breaking change*
- [x] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

Commit:
feat: add config file template

Commit:
feat: add run script

Commit:
docs: update configs template
Add more comments. Remove config that is no longer used.

Commit:
fix: correct imports, comment out checkpoint tidy-up, add docstring
Checkpoint tidy-up doesn't work on S3 buckets. Need to find a method that does.

Commit:
feat: update run script to use boto3 for checkpoint tidy-up
This requires: a) updating the required packages in setup.cfg, b) updating the configs_template.ini so it included additional configs (bucket_name, ssl_file) for the raz client and boto3, c) updating the read_configs method in utils.py so that these configs were read in properly and d) updating main.py so that it contains the new checkpoint tidy-up code and uses the new packages and configs correctly to run this.

Commit:
fix: correct spark session function
Update config spark.shuffle.service.enabled to equal false. This allows the code to run using Spark 3.5.1.

Commit:
docs: update changelog

Commit:
Merge branch 'develop' into feat/run-scripts

Commit:
fix: change pandas version requirements
Cannot specify Pandas >= 2.1.4 as this does not work with Python 3.8. Instead, downgrade to specify Pandas >= 2.0.3, which should work with Python 3.8.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/feat/run-scripts' into feat/run-scripts

Commit:
fix: stop specifying pandas version
Specifying version 2.0.3 still did not work when GitHub Actions were run. Instead, do not specify version until Scalelink stops using Python 3.8. The issue is due to pandas no longer supporting Python 3.8.

Commit:
docs: correct changelog
Configs template was added not updated by this branch.

Commit:
fix: update setup.cfg

Try again...

Commit:
refactor: update tests to reflect spark session and config changes
Update test_get_input_variables to inforporate new config variables bucket_name and ssl_file and remove superseded config variable hdfs_test_path. Update test_create_spark_session so spark.shuffle.service.enabled is always expected to be false.

Commit:
docs: update changelog

Commit:
docs: update to add python 3.11 and remove 3.8

Commit:
ci: update github action that runs unit tests
Update so it now runs them on Python 3.9-3.11.

Commit:
docs: update changelog

mary-cleaton commented:
In draft - waiting til PR #63 is merged as needs to be rebased against that before being ready for QA.

Commit:
ci: remove python 3.9
Python 3.9 ceased to be supported as of 31/10/2025. So, remove it.

Commit:
Merge branch 'develop' into feat/update-python

Commit:
ci: update pull request workflow
The 'test' job in the GitHub Actions pull request workflow was failing with teh following error: ERROR: Failed to build 'gssapi' when getting requirements to build wheel. This was traced to a dependency on krb5-config via a chain originating from raz_client (which we use). Adding this install to the 'test' job fixes the problem. The job refers to 'libkrb5-dev' as this is the name of the krb5-config package in Ubuntu and the GitHub Actions are run using Ubuntu.

danasebo commented on configs_template.ini:
why is this file in this PR?

danasebo commented on configs_template.ini:
```diff
@@ -0,0 +1,28 @@
+spark_session_size =
+
+[filepaths]
+# df1_path, df2_path and df_candidates_path must be to parquet files
```
"[...] must be to parquet [...]" it sounds like a word is missing on this comment line

mary-cleaton replied:
Corrected.

danasebo commented on configs_template.ini:
```diff
@@ -0,0 +1,28 @@
+
+[filepaths]
+# df1_path, df2_path and df_candidates_path must be to parquet files
+# checkpoint_path must be to a unique location as it will be deleted at the end of the run
```
"[...] must be to a unique location [...]" -> "must be a unique location"?

mary-cleaton replied:
Fixed, plus rephrased slightly to make it clearer.

danasebo commented on scalelink/main.py:
```diff
@@ -0,0 +1,124 @@
+
+import boto3
+import raz_client
+from rdsa_utils.cdp.helpers.s3_utils import delete_folder
```
`rdsa_utils` is not included in required packages in `setup.cfg` (and the script does not work unless `rdsa_utils` is installed) - is that intentional?

mary-cleaton replied:
Drat. Not intentional!

I think there's actually several packages missing from the `setup.cfg`. Fixed now. Note, I've commented out PySpark as although it's needed, if you pip install it on our systems it'll break things. We'll need to undo that before publishing - I've put a note in our work record.

danasebo reviewed (changes requested):
overall, all changes look good to me, great work! I tested both the unit tests and the main file, please look at the individual comments for minor requests. once addressed, the PR can be merged.

just to add that I tested that the `main.py` file runs without any issues, as well as the updated unit tests, and I also checked (using grep) that no other unnoticed instances of Python 3.8 / 3.9 are present in the repo.

Commit:
docs: correct typo in configs template

Commit:
docs: rephrase comment in configs template

Commit:
docs: update list of required pacakges
Note: pyspark is commented out as whilst it's required, on our systems installing it using pip breaks things as it's already there.

danasebo commented on setup.cfg:
```diff
@@ -9,18 +9,22 @@ long_description_content_type = text/markdown
 include_package_data = True
 install_requires =
-    pandas
+    boto3 >= 1.34.149
```
silly detail, but could you please update `>= 1.34.149` -> `>=1.34.149` (i.e. removing space after `>=` for consistency? (also for `pandas` & `raz_client`)

mary-cleaton replied:
Done.

danasebo commented on configs_template.ini:
```diff
@@ -0,0 +1,28 @@
+df1_suffix =
+df2_suffix =
+
+[cutpoints]
```
I think it is not clear that the individual cutpoints below need to be defined (i.e. if they are not to be used, they should be `None` rather than kept blank / undefined). I suggest to add this clarification as a comment to the template, if you agree

mary-cleaton replied:
Done.

Commit:
docs: correct formatting error in setup.cfg

Commit:
docs: update configs template to include more instructions re. cutpoints

Commit:
Merge branch 'develop' into feat/update-python

danasebo reviewed (approved):
All looks good to me now, thanks!
